pkg/providers/snow/objects.go

```diff
@@ -80,30 +132,38 @@ func NewMachineTemplateName(new, old *snowv1.AWSSnowMachineTemplate) (string, er
 	return clusterapi.IncrementName(old.GetName())
 }
 
-func UpdateMachineTemplateName(new, old *snowv1.AWSSnowMachineTemplate) error {
-	name, err := NewMachineTemplateName(new, old)
+func NewWorkerMachineTemplateName(newMt, oldMt *snowv1.AWSSnowMachineTemplate, newKct, oldKct *bootstrapv1.KubeadmConfigTemplate) (string, error) {
+	name, err := NewMachineTemplateName(newMt, oldMt)
 	if err != nil {
-		return err
+		return "", err
 	}
-	new.SetName(name)
-	return nil
-}
-
-func WorkerMachineTemplates(ctx context.Context, kubeClient kubernetes.Client, clusterSpec *cluster.Spec) (map[string]*snowv1.AWSSnowMachineTemplate, error) {
-	m := map[string]*snowv1.AWSSnowMachineTemplate{}
 
-	for _, workerNodeGroupConfig := range clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
-		new := SnowMachineTemplate(clusterSpec.SnowMachineConfigs[workerNodeGroupConfig.MachineGroupRef.Name])
+	if oldKct == nil {
+		return name, nil
+	}
 
-		old, err := oldWorkerMachineTemplate(ctx, kubeClient, clusterSpec, workerNodeGroupConfig)
+	if recreateKubeadmConfigTemplateNeeded(newKct, oldKct) {
+		name, err = clusterapi.IncrementName(oldMt.GetName())
 		if err != nil {
-			return nil, err
-		}
-		if err := UpdateMachineTemplateName(new, old); err != nil {
-			return nil, err
+			return "", err
 		}
+	}
+
+	return name, nil
+}
+
+func NewKubeadmConfigTemplateName(new, old *bootstrapv1.KubeadmConfigTemplate) (string, error) {
+	if old == nil {
+		return new.GetName(), nil
+	}
 
-		m[workerNodeGroupConfig.MachineGroupRef.Name] = new
+	if recreateKubeadmConfigTemplateNeeded(new, old) {
+		return clusterapi.IncrementName(old.GetName())
 	}
-	return m, nil
+
+	return old.GetName(), nil
+}
+
+func recreateKubeadmConfigTemplateNeeded(new, old *bootstrapv1.KubeadmConfigTemplate) bool {
+	return !v1alpha1.TaintsSliceEqual(new.Spec.Template.Spec.JoinConfiguration.NodeRegistration.Taints, old.Spec.Template.Spec.JoinConfiguration.NodeRegistration.Taints)
```

Comment:
can we just compare `new.Spec` and `old.Spec`? That would make this a bit more future proof

Reply:
spec field changes should not trigger machine recreate, meaning no new kubeadmconfigtemplate created

Reply:
chatted w folks offline -- we are treating kubeadmconfigtemplate immutable. Refer: https://github.com/aws/eks-anywhere/issues/189#issuecomment-1030421998